wormhole_attester/program/src/attest.rs

```diff
@@ -224,53 +218,84 @@ pub fn attest(ctx: &ExecutionContext, accs: &mut Attest, data: AttestData) -> So
         };
 
-        // Take a mut reference to this price's metadata
-        let state_entry: &mut AttestationState = accs
-            .attestation_state
-            .entries
-            .entry(*price.key)
-            .or_insert(AttestationState {
-                // Use the same value if no state
-                // exists for the symbol, the new value _becomes_ the
-                // last attested trading publish time
-                last_attested_trading_publish_time: last_trading_publish_time,
-            });
+        // Retrieve the timestamp saved during the previous
+        // attestation. Use the new_* value if no existind state is
+        // present on-chain
+        let current_last_attested_trading_publish_time = if state.0 .0.is_initialized() {
+            // Use the existing on-chain value
+            state.0 .0 .1.last_attested_trading_publish_time
+        } else {
+            // Fall back to the new value if the state is not initialized
+            new_last_attested_trading_publish_time
+        };
 
+        // Build an attestatioin struct for this symbol using the just decided current value
         let attestation = PriceAttestation::from_pyth_price_struct(
             Identifier::new(price.key.to_bytes()),
             attestation_time,
-            state_entry.last_attested_trading_publish_time, // Used as last_attested_publish_time
+            current_last_attested_trading_publish_time,
             price_struct,
         );
 
-
-        // update last_attested_publish_time with this price's
-        // publish_time. Yes, it may be redundant for the entry() used
-        // above in the rare first attestation edge case.
-        state_entry.last_attested_trading_publish_time = last_trading_publish_time;
-
-        // The following check is crucial against poorly ordered
-        // account inputs, e.g. [Some(prod1), Some(price1),
-        // Some(prod2), None, None, Some(price)], interpreted by
-        // earlier logic as [(prod1, price1), (prod2, price3)].
-        //
-        // Failing to verify the product/price relationship could lead
-        // to mismatched product/price metadata, which would result in
-        // a false attestation.
-        if attestation.product_id.to_bytes() != product.key.to_bytes() {
-            trace!(&format!(
-                "Price's product_id does not match the pased account (points at {:?} instead)",
-                attestation.product_id
-            ));
-            return Err(ProgramError::InvalidAccountData.into());
+        // Save the new value for the next attestation of this symbol
+        state.0 .0.last_attested_trading_publish_time = new_last_attested_trading_publish_time;
+
+        // handling of last_attested_trading_publish_time ends here
+
+        // Serialize the state to calculate rent/account size adjustments
+        let state_serialized = state.0 .0 .1.try_to_vec()?;
+
+        if state.0 .0.is_initialized() {
```

Comment:
extract a helper function for all of this account allocation logic

Reply:
I think this realloc logic needs to go away. This account is fixed sized and this code is too complex.

Reply:
You're right, this made sense only for the large PDA. Thanks.